### ISSUE TITLE
Toolbar box-shadow mask disappears when keyboard is up

### ISSUE THREAD

**Issue by raineorshine** [cybersemics/em#1785]
Note: This is related to a known problem on mobile Safari with `position: fixed`.

- **Mobile Safari** - Broken ✗ 
- **Xcode Simulator** - Works ✓ 
- **iPhone** - Works ✓
- **Android** - Works ✓

## Steps to Reproduce

Set the caret on a thought to enter edit mode and open the virtual keyboard.

## Current Behavior

The drop shadow on the toolbar disappears when the toolbar is switched from `position: fixed` to `position: absolute`.

https://github.com/cybersemics/em/assets/750276/985acfdb-4813-4bfb-8319-8e92fefb5a00

## Expected Behavior

Drop shadow should always be rendered on the toolbar.

**Comment by msdewitt** [comment]
@raineorshine I'm unable to replicate this error in my local.

Could you provide more information?

![Simulator Screen Recording - iPhone 15 Pro - 2024-11-07 at 12 11 11](https://github.com/user-attachments/assets/3767cc7a-56be-40bb-8d45-27fa5e5f6b8f)

**Comment by raineorshine** [comment]
Platform?

**Comment by msdewitt** [comment]
Safari, Xcode simulator: Iphone 15 pro

**Comment by raineorshine** [comment]
Thanks. It's actually already tested on iPhone Simulator where it works correctly (see original issue description above). We know that it is broken in mobile Safari, but it has not yet been tested on a physical iPhone or Android.

If it's only broken on mobile Safari, then we can focus troubleshooting on that platform, but first we need to test iPhone and Android.

**Comment by msdewitt** [comment]
I'll test it on my android browser.

**Comment by raineorshine** [comment]
@trevinhofmann Could you test this out on a physical iPhone?

**Comment by trevinhofmann** [comment]
@raineorshine - The drop shadow is subtle with the new background displayed on the right side, but I have tested with these results:

✅ Physical Android device (Chrome):

https://github.com/user-attachments/assets/1f1bc8df-2e32-4260-b2e9-059229d04618

❌ Physical iOS device (Safari):

https://github.com/user-attachments/assets/be3a5d0f-85b5-4b5d-8a60-09c5cae4ab98

**Comment by raineorshine** [comment]
Thanks!

In the iOS screenshot, I see a url bar... is that the Capacitor build? Mobile Safari has already been tested, so we're looking to test the Capacitor iOS build on a physical iPhone. 

One way to more clearly see the drop shadow is to scroll down so the toolbar overlaps the first thought. 

## Current

![image](https://github.com/user-attachments/assets/f60fcf5a-c423-424f-9e9d-f33781f73920)

## Expected

![image](https://github.com/user-attachments/assets/be6008ad-7760-4130-9b65-acbb73fcd366)

**Comment by trevinhofmann** [comment]
Ah, my mistake! I misunderstood. I will test this again with Capacitor.

**Comment by trevinhofmann** [comment]
✅ Physical iOS device (Capacitor):

https://github.com/user-attachments/assets/542e2e8e-a857-4a72-87a4-c5b8055faf2a

I'll note that this is affected by #2514, but the drop shadow is still visible and does not disappear when opening the keyboard.

**Comment by raineorshine** [comment]
No, that’s my bad! I definitely didn't communicate that clearly. 

Given that this only affects Mobile Safari, and most iPhone users will use the Capacitor build, this task is fairly low priority. Still, it's worth spending a couple hours on it in case there's an easy fix. 

@msdewitt Are you able to work on this, or should we find a task to replace it with since this is Mobile Safari only?

**Comment by raineorshine** [comment]
I got curious and played around with it until I found a tentative solution in 56da55ecdb. 

Applying the box-shadow directly to the toolbar container and getting ride of the dedicated masking element does not appear to have any regressions, but I will keep an eye on it. I vaguely recall needing an opaque element to mask above the toolbar on overscroll, but I didn't notice any difference when I tested both.

**Comment by msdewitt** [comment]
@raineorshine Awesome!

**Comment by raineorshine** [comment]
Sadly I had to revert my fix because it introduced a regression. Moving the boxShadow to the toolbar container causes it to cover the first few thoughts when the ColorPicker is open.

Up for grabs again!

**Comment by msdewitt** [comment]
@raineorshine After some exploration on this ticket, I don't think your change has caused a regression as the current version in main now still hasthe issue of the toolbar moving upwards when a key is typed while the toolbar is hovering over the top word. I think this bug only happens on IOS currently. Can you confirm this?

**Comment by msdewitt** [comment]
Been looking into this few a few hours now and I can't seem to find the solution. I found the possible problem here: [Link](https://stackoverflow.com/questions/57207030/box-shadow-on-absolute-positioned-div-with-no-content)

It seems to have some issue with the absolute positioned div not having any height to reflect the box shadow for safari. I think this issue is not a simple fix, because we usually just swap a single CSS property when we change the position: fixed to position: absolute. 

@raineorshine and @trevinhofmann, do any of you have some good ideas on how to resolve this minimally based on the link I posted above?

**Comment by raineorshine** [comment]
> Been looking into this few a few hours now and I can't seem to find the solution. I found the possible problem here: [Link](https://stackoverflow.com/questions/57207030/box-shadow-on-absolute-positioned-div-with-no-content)

The StackOverflow question doesn't mention Safari, which makes me think it's unlikely to be the problem here. The absolutely positioned div in **em** has a fixed height anyway.

Note that I can reproduce this on Desktop Safari if I set `isTouch` to true. That might be easier to debug.

---

I played around with it a bit and one possible culprit is `overflowX` that is set in `usePositionedFixed`. On Chrome the box-shadow still works when `overflowX` is set to `hidden`, but on Safari it does not render on the below layer.

https://github.com/cybersemics/em/blob/19f3efc95dcb678c606c9b027c24b0d318c69d45/src/hooks/usePositionFixed.ts#L40

If that is the case, then the task will be to remove `overflowX` or find a suitable replacement without introducing any regressions. There is no comment explaining the purpose of `overflowX`, but it may be able to be identified with git blame.

**Comment by msdewitt** [comment]
@raineorshine You are absolutely correct. There doesn't seem to be a reason set in the gitblame about why that even exists. In fact, when I just keep it as 'visible', the safari issue resolves on its own. I'm thinking I can push up this one line change and merge it, but we keep an eye on it in case there are regressions, like you said. 

I'll submit a pr for review now.

**Comment by raineorshine** [comment]
I still think the git history offers us the best chance of identifying the original purpose. It's admittedly a bit hard to track across multiple refactors and file renames. I don't think we should set it to `visible` without fully understanding its origins, as it must have been added for good reason.

**Comment by msdewitt** [comment]
@raineorshine Here is what I found:
[Blame positionFixed](https://github.com/msdewitt/em/blame/main/src/hooks/usePositionFixed.ts)

**Comment by raineorshine** [comment]
git blame doesn't follow file renames unfortunately. The original change goes much further back. Are you able to investigate a bit further to try to find the origin?

**Comment by msdewitt** [comment]
@raineorshine Yes, I followed it further and I think I found where and why it was added. The reason overflowX is set to hidden when it is set to absolute is because the app starts acting strange with the repositioning, giving a horizontal scroll bar on the bottom of the screen where there should be none.



https://github.com/user-attachments/assets/ddf1e66e-50d6-41aa-9e2f-c8c6ea0d20d2

**Comment by raineorshine** [comment]
Good find, thanks! That would seem to be our main challenge then to fixing this.

**Comment by msdewitt** [comment]
Further information about this issue for reference: [Link](https://stackoverflow.com/questions/36531708/why-does-position-absolute-make-page-to-overflow)
